tls/s2n_early_data_io.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_early_data.h"
+
+#include "tls/s2n_connection.h"
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+int s2n_end_of_early_data_send(struct s2n_connection *conn)
+{
+    if (conn->early_data_expected) {
+        POSIX_GUARD(s2n_stuffer_wipe(&conn->handshake.io));
+        POSIX_BAIL(S2N_ERR_EARLY_DATA_BLOCKED);
+    }
+
+    POSIX_GUARD_RESULT(s2n_connection_set_early_data_state(conn, S2N_END_OF_EARLY_DATA));
+    return S2N_SUCCESS;
+}
+
+int s2n_end_of_early_data_recv(struct s2n_connection *conn)
```

Comment:
Is there a way for a server application to reject early data after receiving a chunk, parsing and validating it, and deciding to reject it? Is that allowed? Or once you've committed to receiving it are you on the hook for the whole thing?

Reply:
Once you've committed to receiving it you're on the hook. You don't have to use it though ¯\_(ツ)_/¯ You could just ignore it all, wait for the EndOfEarlyData, and move on.

Theoretically the server could wait for some early data after the ClientHello, parse that early data, and then change its accept/reject decision on the EncryptedExtensions based on that. But the only thing the accept/reject decision really controls is whether or not the server expects the EndOfEarlyData message, and there are some issues with that:
* The client isn't guaranteed to send any early data, even if it sent the extension indicating it would.
* The client doesn't wait for the server to accept/reject early data before sending more. If it waits for the server's response before continuing, it's not really 0RTT data anymore ;)